jobs/base_model_validator.py

```diff
@@ -0,0 +1,161 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Beam PTransform and DoFn for base model validator functions"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+import re
+
+import apache_beam as beam
+
+from core.domain import cron_services
+from core.platform import models
+from jobs import base_model_validator_errors as errors
+from jobs import jobs_utils
+
+(base_models, user_models) = models.Registry.import_models(
+    [models.NAMES.base_model, models.NAMES.user])
+
+
+MAX_CLOCK_SKEW_SECS = datetime.timedelta(seconds=1)
+
+
+class ValidateModelIdWithRegex(beam.DoFn):
+    """DoFn to validate model ids against a given regex string."""
+
+    def __init__(self, regex_string):
+        """Initializes the ValidateModelIdWithRegex DoFn.
+
+        Args:
+            regex_string: str. A regex pattern for valid ids.
+        """
+        super(ValidateModelIdWithRegex, self).__init__()
+        self.regex = re.compile(regex_string)
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelInvalidIdError. An error class for models with invalid IDs.
+        """
+        model = jobs_utils.clone_model(input_model)
+
+        if not self.regex.match(model.id):
+            yield errors.ModelInvalidIdError(model)
+
+
+class ValidateDeleted(beam.DoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelExpiredError. An error class for expired models.
+        """
+        model = jobs_utils.clone_model(input_model)
+        date_now = datetime.datetime.utcnow()
+
+        expiration_date = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED)
+
+        if model.last_updated < expiration_date:
+            yield errors.ModelExpiredError(model)
+
+
+class ValidateModelTimeFields(beam.DoFn):
+    """DoFn to check whether created_on and last_updated timestamps are
+    valid."""
```

Comment:
Here and below, move closing """ to next line (similar to rest of codebase).

Reply:
Done.

Reply:
So the local lint checks aren't catching this and the only errors being raised by the GitHub lint checks are these:
```
Please fix the errors below:
----------------------------------------
************* Module oppia.jobs.base_model_validator_errors
C: 25, 0: Class name "PERIOD_TO_HARD_DELETE_MODEL_IN_DAYS" doesn't conform to PascalCase naming style 

jobs/base_model_validator_errors.py:82:62: E225 missing whitespace around operator
```

Reply:
Weird, I'm not sure why it's being treated as a class. @DubeySandeep could you please help us debug?

Reply:
pylint ignore might be needed here, see https://github.com/oppia/oppia/pull/7351#discussion_r312731662 for reference.

Reply:
@shubha-rajan The linter errors should have been caught locally also by `pre_commit_linter`. Can you once run the command `python -m scripts.linters.pre_commit_linter --path=jobs/base_model_validator.py` locally and let us know the output log?